### PR TITLE
Refactor perception

### LINKED ISSUE #354: spr is using the python api, not the ros api
rein@athome10 $ grep -ri face_recognizer
challenge_spr_states/detect.py:from openface_ros.face_recognizer import FaceRecognizer
challenge_spr_states/detect.py:        self._face_recognizer = FaceRecognizer(align_path, net_path)
challenge_spr_states/detect.py:        faces = self._face_recognizer.recognize(image)

### LINKED ISSUE #459: [SPR/robot_skills] Implement openface and skybiometry in head
Openface is already implemented. I don't know if the api in the SPR and robot_skills the same are. So check if this is usable and also implement skybiometry in robot_skills

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
Merge together with: tue-robotics/amigo_bringup#81 and tue-robotics/robot_launch_files#20
fixes #459 and fixes #354

### CODE REVIEW

File: robot_skills/src/robot_skills/head.py [hunk 1]
```diff
@@ -22,59 +20,6 @@
 from .util.kdl_conversions import kdlVectorStampedToPointStamped, VectorStamped
 
 
-WavingResult = namedtuple('WavingResult', ['side', 'roi'])
-
-
-class Skeleton(object):
```
Comment (LoyVanBeek) [inline robot_skills/src/robot_skills/head.py]: I see this bit removed but not added again anywhere? We don't need it anymore? @Rayman any ideas?

Comment (MatthijsBurgh) [reply robot_skills/src/robot_skills/head.py]: I couldn't find any usage of it. Any a new skeleton class is used in people_detector

File: robot_skills/src/robot_skills/util/transformations.py [hunk 2]
```diff
@@ -128,6 +128,42 @@ def tf_transform(coordinates, inputframe, outputframe, tf_listener):
     output_coordinates = tf_listener.transformPoint(outputframe, ps, fixed_frame=inputframe)
     return output_coordinates.point
 
+
+def img_recognitions_to_rois(recognitions):
```
Comment (Rayman) [inline robot_skills/src/robot_skills/util/transformations.py]: This has nothing to do with `transformations.py`

File: robot_skills/src/robot_skills/head.py (hunk 1 above)
Comment (Rayman) [reply robot_skills/src/robot_skills/head.py]: This is true, it's all in the people detector package

File: challenge_spr/src/challenge_spr_states/detect.py [hunk 3]
```diff
@@ -57,56 +43,21 @@ def recognize(self, tries):
 
         for i in range(0, tries):
             self.robot.speech.speak(sentences[i % (tries - 1)], block=False)
-            image = self.get_shot()
-            face_rois = self.get_faces(image)
+            self.robot.head.look_at_point(VectorStamped(100, 0, 1.5, self.robot.robot_name + "/base_link"))
+            self.robot.head.wait_for_motion_done()
```
Comment (Rayman) [inline challenge_spr/src/challenge_spr_states/detect.py]: You changed behavior, so this isn't a refactor?

File: robot_skills/src/robot_skills/perception.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+# TU/e
+from rgbd.srv import Project2DTo3D, Project2DTo3DRequest
+
+from .util import msg_constructors as msgs
+from .util.kdl_conversions import kdlVectorStampedToPointStamped, VectorStamped
+
+
+class Perception(RobotPart):
+    def __init__(self, robot_name, tf_listener):
```
Comment (Rayman) [inline robot_skills/src/robot_skills/perception.py]: Did you change behavior of the perception function or just move them?

File: robot_skills/src/robot_skills/util/transformations.py (hunk 2 above)
Comment (MatthijsBurgh) [reply robot_skills/src/robot_skills/util/transformations.py]: It wasn't sure where to put it. I thought this was a better place than adding it as a static method. What do you think?

File: challenge_spr/src/challenge_spr_states/detect.py (hunk 3 above)
Comment (MatthijsBurgh) [reply challenge_spr/src/challenge_spr_states/detect.py]: I didn't plan to change behavior. I see I forgot the sleep. Or did I forget more?

File: robot_skills/src/robot_skills/head.py (hunk 1 above)
Comment (LoyVanBeek) [reply robot_skills/src/robot_skills/head.py]: OK then

File: robot_skills/src/robot_skills/util/transformations.py (hunk 2 above)
Comment (LoyVanBeek) [reply robot_skills/src/robot_skills/util/transformations.py]: As said before: a separate module besides transformations.py, eg. image_operations.py

File: robot_skills/src/robot_skills/util/image_operations.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    cutouts = []
+    cv2_image = cvbridge.imgmsg_to_cv2(image, 'bgr8')
+    for roi in rois:
+        cv2_cutout = cv2_image[roi.y_offset:roi.y_offset + roi.height, roi.x_offset:roi.x_offset + roi.width]
+        cutout = cvbridge.cv2_to_imgmsg(cv2_cutout, 'bgr8')
+        cutouts.append(cutout)
+    return cutouts
+
+
+if __name__ == '__main__':
```
Comment (LoyVanBeek) [inline robot_skills/src/robot_skills/util/image_operations.py]: Useless without actual doctests, but it's a start.

File: robot_skills/src/robot_skills/util/image_operations.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    import doctest
+    doctest.testmod()
```
Comment (LoyVanBeek) [inline robot_skills/src/robot_skills/util/image_operations.py]: Useless without actual doctests, but a nice start

**Comment by LoyVanBeek** [comment]
What is the impact on the challenges? Is the refactor also reflected on those?

**Comment by MatthijsBurgh** [comment]
checked the whole trunk, so everything should keep working.

**Comment by reinzor** [comment]
Why are we even working on this? What is the issue? I think in general we should work on issues instead of doing work without any explicit motivation.

**Comment by MatthijsBurgh** [comment]
I didn't have a task yet for object recognition. Plus, I do some small side projects all the time.

File: robot_skills/src/robot_skills/mockbot.py [hunk 7]
```diff
@@ -193,8 +198,6 @@ def __init__(self, *args, **kwargs):
 
         self._person_names = []
 
-        self.learn_person = lambda name: True
```
Comment (LoyVanBeek) [inline robot_skills/src/robot_skills/mockbot.py]: Removed but no-where added back

Comment (MatthijsBurgh) [reply robot_skills/src/robot_skills/mockbot.py]: This function is not part of the ED robot_part anymore. I did some cleaning of the mockbot, But the mockbot is not in sync with the rest at all.

File: robot_skills/package.xml [hunk 8]
```diff
@@ -58,12 +58,6 @@
   <build_depend>cb_planner_msgs_srvs</build_depend>
   <run_depend>cb_planner_msgs_srvs</run_depend>
 
-  <build_depend>octomap_msgs</build_depend>
-  <run_depend>octomap_msgs</run_depend>
-
-  <build_depend>cb_planner_msgs_srvs</build_depend>
-  <run_depend>cb_planner_msgs_srvs</run_depend>
-
```
Comment (Rayman) [inline robot_skills/package.xml]: You are doing too much stuff changes in one PR. It makes it really hard to review. Can you please make PR's based on logical chunks of work?

**Comment by Rayman** [comment]
@SamAlexandrov should be involved as well because of changes in his package

File: robot_skills/package.xml (hunk 8 above)
Comment (MatthijsBurgh) [reply robot_skills/package.xml]: You are right. Do you want me to create separate PR's now? Or only for future PR's

Comment (LoyVanBeek) [reply robot_skills/package.xml]: For now, id' revert the removal of this dep and make an issue/PR of cleanup of the dependencies. Which has low prio.

**Pull request merged**